src/Parsers/Access/ParserExecuteAsQuery.cpp

```diff
@@ -0,0 +1,63 @@
+#include <Parsers/Access/ParserExecuteAsQuery.h>
+#include <Parsers/Access/ASTRolesOrUsersSet.h>
+#include <Parsers/Access/ASTExecuteAsQuery.h>
+#include <Parsers/Access/ParserRolesOrUsersSet.h>
+#include <Parsers/ParserSelectWithUnionQuery.h>
+#include <Parsers/ASTSelectWithUnionQuery.h>
+#include <Parsers/CommonParsers.h>
+
+
+namespace DB
+{
+namespace
+{
+    bool parseTargetUser(IParserBase::Pos & pos, Expected & expected, std::shared_ptr<ASTRolesOrUsersSet> & targetuser)
+    {
+        return IParserBase::wrapParseImpl(pos, [&]
+        {
+            ASTPtr ast;
+            ParserRolesOrUsersSet user_p;
+            user_p.allowUsers();
+            if (!user_p.parse(pos, ast, expected))
+                return false;
+
+            targetuser = typeid_cast<std::shared_ptr<ASTRolesOrUsersSet>>(ast);
+            targetuser->allow_roles = false;
+            return true;
+        });
+    }
+}
+
+
+bool ParserExecuteAsQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+{
+    std::shared_ptr<ASTRolesOrUsersSet> targetuser;
+
+    if (ParserKeyword{Keyword::EXECUTE_AS}.ignore(pos, expected))
+    {
+        if (!parseTargetUser(pos, expected, targetuser))
+            return false;
+    }
+    else
+        return false;
+
+    auto query = std::make_shared<ASTExecuteAsQuery>();
+    node = query;
+
+    query->targetuser = targetuser;
+
+    /// support 1) EXECUTE AS <user1>  2) EXECUTE AS <user1> SELECT ...
```

Comment:
It makes sense to support any query after `EXECUTE AS <user>`, not just `SELECT`.

For example,
```
EXECUTE AS <user1> CREATE TABLE ...
```

Reply:
I added support for that.